RHDEVDOCS-4073 - Document on binding workloads using the label selector feature as per the spec 1.0 (#46026)
openshift/openshift-docs: modules/sbo-advanced-binding-options.adoc

= Advanced binding options

[role="_abstract"]
You can define the `ServiceBinding` custom resource (CR) to use the following advanced binding options:

* Changing binding names: This option is available only for the `binding.operators.coreos.com` API group.
* Composing custom binding data: This option is available only for the `binding.operators.coreos.com` API group.
* Binding workloads using label selectors: This option is available for both the `binding.operators.coreos.com` and `servicebinding.io` API groups.

[id="changing-binding-names_{context}"]
== Changing the binding names before projecting them into the workload

> You can specify the rules to change the binding names in the .spec.namingStrategy attribute of the `ServiceBinding` CR. For example, consider a Spring PetClinic sample application that connects to the PostgreSQL database. In this case, the PostgreSQL database service exposes the host and port fields of the database to use for binding. The Spring PetClinic sample application can access this exposed binding data through the binding names.

Reviewer: Check with the SME whether these field names are allcaps or not. For example "the `HOST` and `PORT` fields"

> You can specify the rules to change the binding names in the .spec.namingStrategy attribute of the `ServiceBinding` CR. For example, consider a Spring PetClinic sample application that connects to the PostgreSQL database. In this case, the PostgreSQL database service exposes the host and port fields of the database to use for binding. The Spring PetClinic sample application can access this exposed binding data through the binding names.

Author: Fixed

<1> Name of the backing service resource.
<2> Optional identifier.
<3> Generated JSON name that is to be projected as the file content or environment value. The JSON name contains the attributes of the backing service custom resource.

> <4> Generated JSON value that is to be projected as the file content or environment value. The JSON value contains the attributes of the backing service custom resource.

Reviewer: The "that is to be projected" wording sounds strange. It contains both future tense and passive voice.

> <4> Generated JSON value that is to be projected as the file content or environment value. The JSON value contains the attributes of the backing service custom resource.

Author: This is the version that I prefer and to the point:

"The JSON value that the {servicebinding-title} generates. The {servicebinding-title} projects this JSON value as a file or environment variable. The JSON value contains the attributes from the specified field of the backing service custom resource."

> <4> Generated JSON value that is to be projected as the file content or environment value. The JSON value contains the attributes of the backing service custom resource.

Author: "The JSON name that the {servicebinding-title} generates. The {servicebinding-title} projects this JSON name as a file or environment variable. The JSON name contains the attributes from your specified field of the backing service custom resource."

[id="binding-workloads-using-a-label-selector_{context}"]
== Binding workloads using a label selector

> You can use a label selector to specify the workload that is being bound. If you declare a service binding using the label selectors to pick up workloads, the {servicebinding-title} periodically attempts to find and bind new workloads that match the given label selector.

Reviewer: How about:
```suggestion
You can use a label selector to specify the workload to bind. If you declare a service binding using the label selectors to pick up workloads, the {servicebinding-title} periodically attempts to find and bind new workloads that match the given label selector.
```
This suggestion incorporates the following guidelines:
- Avoid passive voice.

> For example, you may want to bind a service to every `Deployment` in a namespace with the `environment: production` label set in the `application` field. By setting an appropriate label selector, you can enable the {servicebinding-title} to bind each of these workloads with one `ServiceBinding` resource.

Reviewer: ```suggestion
For example, to bind a service to every `Deployment` in a namespace, you set the `environment: production` label as the value of the `application` field. By setting an appropriate label selector, you enable the {servicebinding-title} to bind each of these workloads with one `ServiceBinding` resource.
```

> For example, you may want to bind a service to every `Deployment` in a namespace with the `environment: production` label set in the `application` field. By setting an appropriate label selector, you can enable the {servicebinding-title} to bind each of these workloads with one `ServiceBinding` resource.

Reviewer: Or...
```suggestion
For example, you can bind a service to every `Deployment` in a namespace by using the `environment: production` label as the value of the `application` field. Setting the appropriate label selector enables the {servicebinding-title} to bind each of these workloads with one `ServiceBinding` resource.
```

> For example, you may want to bind a service to every `Deployment` in a namespace with the `environment: production` label set in the `application` field. By setting an appropriate label selector, you can enable the {servicebinding-title} to bind each of these workloads with one `ServiceBinding` resource.

Author: For example, as a cluster administrator, you can bind a service to every `Deployment` in a namespace with the `environment: production` label by setting an appropriate `labelSelector` field in the `ServiceBinding` CR.  This enables the {servicebinding-title} to bind each of these workloads with one `ServiceBinding` CR.

> For example, you may want to bind a service to every `Deployment` in a namespace with the `environment: production` label set in the `application` field. By setting an appropriate label selector, you can enable the {servicebinding-title} to bind each of these workloads with one `ServiceBinding` resource.

Author: Fixed

> .Example of `ServiceBinding` CR in the `binding.operators.coreos.com/v1alpha1` API:

Author: Fixed

[source,yaml]
----
apiVersion: binding.operators.coreos.com/v1alpha1
kind: ServiceBinding
metadata:
  name: multi-application-binding
  namespace: service-binding-demo
spec:
  application:
    labelSelector: <1>
      matchLabels:
        environment: production
    group: apps
    version: v1
    resource: deployments
  services:
    group: ""
    version: v1
    kind: Secret
    name: super-secret-data
----
<1> Specifies the workload that is being bound.

> .Example of `ServiceBinding` CR in the `servicebinding.io` API:

Reviewer: ```suggestion
.Example `ServiceBinding` CR in the `servicebinding.io` API
```

Author: Fixed

[source,yaml]
----
apiVersion: servicebindings.io/v1alpha3
kind: ServiceBinding
metadata:
  name: multi-application-binding
  namespace: service-binding-demo
spec:
  workload:
    selector: <1>
      matchLabels:
        environment: production
    apiVersion: app/v1
    kind: Deployment
  service:
    apiVersion: v1
    kind: Secret
    name: super-secret-data
----
<1> Specifies the workload that is being bound.

[IMPORTANT]
====

> Currently, it is forbidden to attempt a binding with the following fields defined:

Author: Fixed

* Both `name` and `labelSelector` in the `binding.operators.coreos.com/v1alpha1` API
* Both `name` and `selector` in the `servicebinding.io` API (Spec API)

> Any attempt to do so will result in an error.

Reviewer: Rephrase to use the present tense.

> * Both `name` and `labelSelector` in the `binding.operators.coreos.com/v1alpha1` API
> * Both `name` and `selector` in the `servicebinding.io` API (Spec API)
> 
> Any attempt to do so will result in an error.

Reviewer: ```suggestion
* The `name` and `labelSelector` fields in the `binding.operators.coreos.com/v1alpha1` API
* The `name` and `selector` fields in the `servicebinding.io` API (Spec API)
```

Reviewer: If you keep line 185, please rephrase it to avoid using future tense.

Author: Fixed
